### ISSUE TITLE
[FARGATE] [request]: Image labels are not visible on the metadata endpoint

### ISSUE THREAD

**Issue by libc** [aws/containers-roadmap#757]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
ECS Task definition supports `dockerLabel` map, that adds labels to the container.  These labels are visible in the container, if queried from the metadata server (`curl -s 169.254.170.2/v2/metadata | jq .Labels`).

`Dockerfile` supports `LABEL` statements that add labels to the image, when running locally with `docker run`, these labels are assigned to the container (and visible with `docker inspect $container_id`). When running in fargate, these labels are not visible on the metadata server (when queried by `curl -s 169.254.170.2/v2/metadata | jq .Labels`)

**Which service(s) is this request for?**
Fargate. ECS might also be affected by this

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
Trying to add additional metadata (git-sha + build version) to metrics collected by 3rd party agent. For example Datadog supports `DD_DOCKER_LABELS_AS_TAGS` option, which allows tagging everything coming out from the agent.

**Are you currently working around this issue?**
Work around would be updating task definition

**Additional context**
I am not the only person to run into this: [this stack exchange question](https://devops.stackexchange.com/questions/6825/docker-image-labels-in-fargate-are-not-applied-to-containers).

**Comment by nicholasgcoles** [comment]
@libc based on the task definition metadata URL you are using it seems that `Labels` is nested in `Containers`. This can be seen in the [docs](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint-v2.html).

If however, you were querying a single container endpoint, you could possibly find it in your environment variables as `$ECS_CONTAINER_METADATA_URI` and then your query should work. I.e.
```
curl -s $ECS_CONTAINER_METADATA_URI | jq .Labels
```

Lastly, I have discovered that these labels are available by default:
```
com.amazonaws.ecs.cluster
com.amazonaws.ecs.container-name
com.amazonaws.ecs.task-arn
com.amazonaws.ecs.task-definition-family
com.amazonaws.ecs.task-definition-version
```

I hope this helps?

**Comment by mchristiansen** [comment]
@nicholasgcoles You are correct that `curl -s $ECS_CONTAINER_METADATA_URI | jq .Labels` is correct for querying the container labels, but the issue @libc is calling out is still a problem because the container labels do not include the image labels. It seems Docker copies the image labels to the container on creation where Fargate does not.

**Comment by hugoivr** [comment]
@libc have you been able to find anything else related to this issue?

**Comment by c-ameron** [comment]
Hi, this has been open for more than 3 years now.

Has there been any more developments or progress in this being picked up?

Thanks

**Comment by bhsp** [comment]
It's common for application & infrastructure monitoring services to use labels (org.label-schema.* and org.opencontainers.*) for correlation. Fargate promoting image labels to container labels as Docker does is a needed feature. Labels org.opencontainers.image.source & org.opencontainers.image.revision should be built at the image and promoted to the container.  Thanks!
